src/coreclr/jit/morph.cpp

```diff
@@ -13969,6 +13977,143 @@ GenTree* Compiler::fgOptimizeBitwiseAnd(GenTreeOp* andOp)
     return nullptr;
 }
 
+//------------------------------------------------------------------------
+// fgOptimizeRelationalComparisonWithCasts: Recognizes comparisons against
+//   various cast operands and tries to remove them. E.g.:
+//
+//   *  GE        int
+//   +--*  CAST      long <- ulong <- uint
+//   |  \--*  X         int
+//   \--*  CNS_INT   long
+//
+//   to:
+//
+//   *  GE_un     int
+//   +--*  X         int
+//   \--*  CNS_INT   int
+//
+//   same for:
+//
+//   *  GE        int
+//   +--*  CAST      long <- ulong <- uint
+//   |  \--*  X         int
+//   \--*  CAST      long <- [u]long <- int
+//      \--*  ARR_LEN   int
+//
+// Arguments:
+//   tree - the GT_LE/GT_LT/GT_GE/GT_GT tree to morph.
+//
+// Return Value:
+//   Returns the same tree where operands might have narrower types
+//
+GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* tree)
+{
+    assert(tree->OperIs(GT_LE, GT_LT, GT_GE, GT_GT));
+
+    GenTree* castedOp        = tree->gtGetOp1();
+    GenTree* knownPositiveOp = tree->gtGetOp2();
+
+    // Caller is expected to call this function only if we have CAST nodes
+    assert(castedOp->OperIs(GT_CAST) || knownPositiveOp->OperIs(GT_CAST));
+
+    if (optValnumCSE_phase)
+    {
+        // We're going to modify all of them
+        return tree;
+    }
+
+    if (!castedOp->TypeIs(TYP_LONG))
+    {
+        // We can extend this logic to handle small types as well, but currently it's done mostly to
+        // assist range check elimination
+        return tree;
+    }
+
+    bool knownPositiveIsOp2;
+    if (knownPositiveOp->IsIntegralConst() ||
+        ((knownPositiveOp->OperIs(GT_CAST) && knownPositiveOp->gtGetOp1()->OperIs(GT_ARR_LENGTH))))
+    {
+        // op2 is either a LONG constant or (T)ARR_LENGTH
+        knownPositiveIsOp2 = true;
+    }
+    else
+    {
+        // op1 is either a LONG constant (yes, it's pretty normal for relops)
+        // or (T)ARR_LENGTH
+        castedOp           = tree->gtGetOp2();
+        knownPositiveOp    = tree->gtGetOp1();
+        knownPositiveIsOp2 = false;
+    }
+
+    if (castedOp->OperIs(GT_CAST) && varTypeIsLong(castedOp->CastToType()) && castedOp->gtGetOp1()->TypeIs(TYP_INT) &&
```

Comment:
```suggestion
    if (castedOp->OperIs(GT_CAST) && varTypeIsLong(castedOp->CastToType()) && castedOp->AsCast()->CastOp()->TypeIs(TYP_INT) &&
```
(Also the `TypeIs(TYP_INT)` condition could be relaxed to `genActualTypeIsInt`)